flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java

```diff
@@ -0,0 +1,172 @@
+
+abstract class AbstractStateChangeLogger<Key, Value, Ns> implements StateChangeLogger<Value, Ns> {
+    protected final StateChangelogWriter<?> stateChangelogWriter;
+    protected final InternalKeyContext<Key> keyContext;
+
+    public AbstractStateChangeLogger(
+            StateChangelogWriter<?> stateChangelogWriter, InternalKeyContext<Key> keyContext) {
+        this.stateChangelogWriter = checkNotNull(stateChangelogWriter);
+        this.keyContext = checkNotNull(keyContext);
+    }
+
+    @Override
+    public void valueUpdated(Value newValue, Ns ns) throws IOException {
+        if (newValue == null) {
+            valueCleared(ns);
+        } else {
+            log(SET, out -> serializeValue(newValue, out), ns);
+        }
+    }
+
+    @Override
+    public void valueUpdatedInternal(Value newValue, Ns ns) throws IOException {
+        if (newValue == null) {
+            valueCleared(ns);
+        } else {
+            log(SET_INTERNAL, out -> serializeValue(newValue, out), ns);
+        }
+    }
+
+    protected abstract void serializeValue(Value value, DataOutputViewStreamWrapper out)
+            throws IOException;
+
+    @Override
+    public void valueAdded(Value addedValue, Ns ns) throws IOException {
+        log(ADD, out -> serializeValue(addedValue, out), ns);
+    }
+
+    @Override
+    public void valueCleared(Ns ns) throws IOException {
+        log(CLEAR, out -> {}, ns);
+    }
+
+    @Override
+    public void valueElementAdded(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataSerializer, Ns ns)
+            throws IOException {
+        log(ADD_ELEMENT, dataSerializer, ns);
+    }
+
+    @Override
+    public void valueElementAddedOrUpdated(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataSerializer, Ns ns)
+            throws IOException {
+        log(ADD_OR_UPDATE_ELEMENT, dataSerializer, ns);
+    }
+
+    @Override
+    public void valueElementRemoved(
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataSerializer, Ns ns)
+            throws IOException {
+        log(REMOVE_ELEMENT, dataSerializer, ns);
+    }
+
+    protected void log(
+            StateChangeOperation op,
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter,
+            Ns ns)
+            throws IOException {
+        // todo: log metadata (FLINK-22808)
+        stateChangelogWriter.append(
+                keyContext.getCurrentKeyGroupIndex(), serialize(op, ns, dataWriter));
+    }
+
+    private byte[] serialize(
+            StateChangeOperation op,
+            Ns ns,
+            ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter)
+            throws IOException {
+        return serializeRaw(
+                wrapper -> {
+                    wrapper.writeByte(op.code);
+                    serializeScope(ns, wrapper);
+                    dataWriter.accept(wrapper);
```

Comment:
Since some `dataWriter` could be empty as `out -> {}`, I suppose it might help improve a bit performance by passing a `null` dataWriter here and check whether this is null to avoid creating a new lambda `out -> {}` here.